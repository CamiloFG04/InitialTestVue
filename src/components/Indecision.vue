<template>
  <img v-if="img" :src="img" alt="bg">
  <div class="bg-dark"></div>
  <div class="indecision-container">
    <input type="text" placeholder="Make me a question" v-model="question">
    <p>Remember end with a question mark (?)</p>

    <div v-if="isValidQuestion">
        <h2>{{question}}</h2>
        <h1>{{answer}}</h1>
    </div>
  </div>
</template>

<script>
export default {
    data(){
        return{
            question: null,
            answer: null,
            img: null,
            isValidQuestion: false
        }
    },
    methods:{
        async getAnswer(){
            try {
                this.answer = 'Thinking...'
                const {answer, image} = await fetch('https://yesno.wtf/api').then(r =>r.json())
                this.answer = answer;
                this.img = image;
            } catch (error) {
                console.log('error:',error);
                this.answer = 'Error Server 500'
                this.img = null
            }
        }
    },
    watch:{
        question(value,oldValue){
            this.isValidQuestion = false
            console.log({value});
            if(!value.includes('?')) return;
            this.isValidQuestion = true
            this.getAnswer()
            // TODO: Make http request
        }
    }
}
</script>

<style scoped>

    img, .bg-dark {
        height: 100vh;
        left: 0px;
        max-height: 100%;
        max-width: 100%;
        position: fixed;
        top: 0px;
        width: 100vw;
    }

    .bg-dark {
        background-color: rgba(0, 0, 0, 0.4);
    }

    .indecision-container {
        position: relative;
        z-index: 99;
    }
    
    input {
        width: 250px;
        padding: 10px 15px;
        border-radius: 5px;
        border: none;
        margin-bottom: 5px;
    }
    input:focus {
        outline: none;
    }

    p {
        color: white;
        font-size: 20px;
        margin-top: 0px;
    }

    h1, h2 {
        color: white;
    }
    
    h2 {
        margin-top: 150px;
    }

</style>