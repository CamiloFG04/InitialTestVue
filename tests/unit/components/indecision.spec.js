import { shallowMount } from "@vue/test-utils";
import Indecision from "@/components/Indecision";

describe('Indecicion Component', () => {
    let wrapper;
    let clgSpy;
    let input;

    global.fetch = jest.fn(()=> Promise.resolve({
        json: ()=> Promise.resolve({
            answer: 'yes',
            forced: false,
            image: 'https://yesno.wtf/assets/yes/2.gif'
        })
    }))

    beforeEach(() => {
        wrapper = shallowMount(Indecision)
        clgSpy = jest.spyOn(console,'log')
        input = wrapper.find('input');
        jest.clearAllMocks()
    });

    test('should be match with snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot()
    });

    test('writing to the input should not trigger anything just console.log', async () => {
        const getAnswerSpy = jest.spyOn(wrapper.vm,'getAnswer')
        await input.setValue('Hola')
        expect(clgSpy).toHaveBeenCalled();
        expect(getAnswerSpy).not.toHaveBeenCalled();
    });

    test('writing "?" must trigger getAnswer ', async() => {
        const getAnswerSpy = jest.spyOn(wrapper.vm,'getAnswer')
        await input.setValue('Hola?')
        expect(getAnswerSpy).toHaveBeenCalled();
    });

    test('tests in getAnswer', async () => {
        await wrapper.vm.getAnswer()
        const img = wrapper.find('img').exists()

        expect(img).toBeTruthy()

        expect(wrapper.vm.img).toBe('https://yesno.wtf/assets/yes/2.gif')

        expect(wrapper.vm.answer).toBe('yes')
    });

    test('test in getAnswer - failure in API', async() => {
        fetch.mockImplementationOnce(()=>Promise.reject('Error'))
        await wrapper.vm.getAnswer()

        const img = wrapper.find('img').exists()
        expect(img).toBeFalsy()
        expect(wrapper.vm.answer).toBe('Error Server 500')
    });
});