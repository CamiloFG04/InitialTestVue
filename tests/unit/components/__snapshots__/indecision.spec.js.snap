// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecicion Component should be match with snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Make me a question">
  <p>Remember end with a question mark (?)</p>
  <!--v-if-->
</div>
`;
